examples/ssr-data-router/src/entry.client.tsx

```diff
@@ -4,11 +4,17 @@ import { createBrowserRouter, RouterProvider } from "react-router-dom";
 
 import { routes } from "./App";
 
-let router = createBrowserRouter(routes);
-
-ReactDOM.hydrateRoot(
-  document.getElementById("app"),
-  <React.StrictMode>
-    <RouterProvider router={router} fallbackElement={null} />
-  </React.StrictMode>
-);
+createBrowserRouter(routes, {
+  // If you're using lazy route modules and you haven't yet preloaded them onto
+  // routes, then you'll need to wait for the router to be initialized before
+  // hydrating, since it will have initial data to hydrate but it won't yet have
+  // any router elements to render.
+  onInitialize({ router }) {
+    ReactDOM.hydrateRoot(
+      document.getElementById("app")!,
+      <React.StrictMode>
+        <RouterProvider router={router} fallbackElement={null} />
+      </React.StrictMode>
+    );
+  },
+});
```

Comment:
@mjackson  @ryanflorence Curious what you think of this API?  It's basically a convenience API for:

```js
let router = createBrowserRouter(routes);

if (router.state.initialized) {
  cb();
} else {
  let unsub = router.subscribe(state => {
    if (state.initialized) {
      unsub();
      cb();
    }
  });
}
```

It's also generally only really needed for SSR + `route.lazy()` where your SSR stack hasn't pre-loaded the lazy route modules and you can't just rely on `<RouteProvider fallbackElement>` due to the SSR'd contents.  I might be leaning towards this feeling like a small enough case that we could just document the edge case + manual solution and leave the API surface alone.

Reply:
If you're doing SSR with lazy routes, you really only need to initialize the portion of the route tree that matches the location you wanna render, right?

Could we do something like:

```ts
let router = createStaticRouter(routes);
await router.initialize();
// ... render(router)
```

??

Reply:
This is already the case via the `context` arg. Internally we resolve lazy routes for the current matches based on the request, with the lazy-loaded route properties being updated in place before being passed to `createStaticRouter`:

```ts
let { query, dataRoutes } = createStaticHandler(routes);
let context = await query(request);

let router = createStaticRouter(dataRoutes, context);
```

We're referring to the client hydration step where typically all elements are synchronously available for rendering, but this isn't true when they need to be lazy loaded first, so we need to delay hydration until the router is initialized.

@mjackson For now I've added a strawman implementation closer to your suggestion which looks like this:

```ts
let router = await createBrowserRouter(routes).ready();
```

The name could definitely use some work, although tbh I don't mind it. I went with `ready: () => Promise<Router>` because `initialize` is already taken and is currently synchronous.

(Side note, I was also guided towards swapping out our `onInitialize` callback for the `ready()` async function because I was adding tests and our previous API made this impractical)

Reply:
I still find myself on the fence as to whether we need this DX shorthand instead of manual `subscribe()` call given it's not the common use-case - but I'm on board with the `ready` API if we want to include it 👍